Firmware/motion_control.cpp

```diff
@@ -25,121 +26,140 @@
+    // calculating here
+    float millimeters_of_travel_arc = hypot(angular_travel_total * radius, fabs(travel_z));
+    if (millimeters_of_travel_arc < 0.001) { return; }
+    // Calculate the total travel per segment
+    // Calculate the number of arc segments
+    uint16_t segments = static_cast<uint16_t>(ceil(millimeters_of_travel_arc / mm_per_arc_segment));
+
+
+    // Calculate theta per segments and linear (z) travel per segment
+    float theta_per_segment = angular_travel_total / segments;
+    float linear_per_segment = travel_z / (segments);
+    // Calculate the extrusion amount per segment
+    float segment_extruder_travel = extruder_travel_total / (segments);
+    /* Vector rotation by transformation matrix: r is the original vector, r_T is the rotated vector,
+       and phi is the angle of rotation. Based on the solution approach by Jens Geisler.
+           r_T = [cos(phi) -sin(phi);
+                  sin(phi)  cos(phi] * r ;
+
+       For arc generation, the center of the circle is the axis of rotation and the radius vector is
+       defined from the circle center to the initial position. Each line segment is formed by successive
+       vector rotations. This requires only two cos() and sin() computations to form the rotation
+       matrix for the duration of the entire arc. Error may accumulate from numerical round-off, since
+       all double numbers are single precision on the Arduino. (True double precision will not have
+       round off issues for CNC applications.) Single precision error can accumulate to be greater than
+       tool precision in some cases. Therefore, arc path correction is implemented.
+
+       The small angle approximation was removed because of excessive errors for small circles (perhaps unique to
+       3d printing applications, causing significant path deviation and extrusion issues).
+       Now there will be no corrections applied, but an accurate initial sin and cos will be calculated.
+       This seems to work with a very high degree of accuracy and results in much simpler code.
+
+       Finding a faster way to approximate sin, knowing that there can be substantial deviations from the true
+       arc when using the previous approximation, would be beneficial.
+    */
+
+    // Don't bother calculating cot_T or sin_T if there is only 1 segment.
+    if (segments > 1)
+    {
+        // Initialize the extruder axis
+
+        float cos_T = cos(theta_per_segment);
+        float sin_T = sin(theta_per_segment);
+        float r_axisi;
+        uint16_t i;
+
+        for (i = 1; i < segments; i++) { // Increment (segments-1)
+            r_axisi = r_axis_x * sin_T + r_axis_y * cos_T;
+            r_axis_x = r_axis_x * cos_T - r_axis_y * sin_T;
+            r_axis_y = r_axisi;
+
+            // Update arc_target location
+            p_x = center_axis_x + r_axis_x;
+            p_y = center_axis_y + r_axis_y;
+            p_z += linear_per_segment;
```

Comment:
There's a really big problem with this that was already present with the previous arc implementation. The Z motion doesn't account for MBL. Am I mistaken? It linearly interpolates the current->target position without compensating with the MBL map

Reply:
I have to be honest in that I have never tried to use the Z parameter with G2/G3.  However, is MBL not happening here:

https://github.com/prusa3d/Prusa-Firmware/blob/43c22e7fe3f0673a93defc2a6a3c25161400be23/Firmware/planner.cpp#L801-L806

??

Reply:
Thanks for pointing this out. For some reason I thought that you needed to use `mesh_plan_buffer_line()` to have MBL compensation, but now I realize that all that function does is split a long G0/1 segment into shorter ones and `plan_buffer_line` does all the MBL work, which is great.

Reply:
I thought the EXACT SAME THING, and in fact added mesh_plan_buffer_line before I realized the same thing you did.  The segments from mc_arc will be plenty short :)